jax/_src/custom_call.py

```diff
@@ -0,0 +1,117 @@
+  row-major layouts.
+
+  If keyword arguments are passed when lowering, these are treated as attributes,
+  and passed to the custom call using the `backend_config` interface available
+  when `api_version=4`. This behavior can be overridden by explicitly passing a
+  value for `backend_config`.
+
+  Args:
+    call_target_name: The name of the custom call target.
+    api_version: The ABI contract version for the custom call.
+    backend_config: Configuration data for the custom call. If not provided,
+      this will be constructed using any keyword arguments passed when lowering.
+    result_types: The MLIR types of the results. By default, the types of the
+      output avals will be used.
+    result_shapes: Tensors that represent the result shapes, to be used when
+      the results have dynamic shapes. By default, this will be constructed
+      automatically based on the shapes of the output avals.
+    operand_layouts: A sequence of layouts (dimension orders) for each operand.
+      By default, the operands are assumed to be row-major.
+    result_layouts: A sequence of layouts (dimension orders) for each result.
+      By default, the results are assumed to be row-major.
+    lowering_args: Any other arguments to :func:`mlir.custom_call` will also be
+      passed through if provided as extra arguments to this function.
+  """
+
+  def _lowering(
+    ctx: mlir.LoweringRuleContext, *operands: ir.Value, **params: Any
+  ) -> Sequence[ir.Value | Sequence[ir.Value]]:
+    kwargs = dict(lowering_args)
+    if params and backend_config is None:
+      assert api_version >= 4
+      kwargs["backend_config"] = {k: _ir_attribute(v) for k, v in params.items()}
+    if result_types is None:
+      kwargs["result_types"] = [mlir.aval_to_ir_type(aval) for aval in ctx.avals_out]
+
+    # TODO(dfm): This is a common pattern for supporting dynamic shapes, but I
+    # haven't tested this implementation yet.
+    if result_shapes is None:
+      if any(not core.is_constant_shape(aval_out.shape) for aval_out in ctx.avals_out):
+        kwargs["result_shapes"] = [
+            mlir.shape_tensor(mlir.eval_dynamic_shape(ctx, aval_out.shape))
+            for aval_out in ctx.avals_out]
+
+    if operand_layouts is None:
+      kwargs["operand_layouts"] = _default_layouts(aval.shape for aval in ctx.avals_in)  # pytype: disable=attribute-error
+    if result_layouts is None:
+      kwargs["result_layouts"] = _default_layouts(aval.shape for aval in ctx.avals_out)
+
+    return mlir.custom_call(
+      call_target_name, operands=operands, api_version=api_version, **kwargs
+    ).results  # type: ignore
+
+  return _lowering
+
+
+def _default_layouts(shapes: Iterable[tuple[int]]):
+  return [list(reversed(range(len(shape)))) for shape in shapes]
+
+
+def _ir_attribute(obj: Any) -> ir.Attribute:
+  # TODO(dfm): Similar functions exist in Pallas and Mosaic GPU. Perhaps these
+  # could be consolidated into mlir or similar.
```

Comment:
Definitely!

Reply:
I'm inclined to keep this here for now to keep this PR compact, then we could move this function later if it would be useful to consolidate.